Fix incorrect net sales in analytics product reports for partial and full refunds

ianlin opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Part of the epic issue #50480, this PR fixes current WC analytics inconsistent net sales for **partial** and **full** refunds.
Those fixes were tested and reviewed in #48768 and #51262, ~so technically this umbrella PR doesn't require a lot of detailed testing~.

**Update 5th March 2025**

After a long back and forth reviews and fixes, this PR actually solves the problems not specifically to the inconsistent net sales between product and order reports, but the product reports itself by making the table `wp_wc_order_product_lookup` display every amount correctly when the partial and full refunds happen. Before those fixes, there were no record in the table `wp_wc_order_product_lookup` when refunds.

✏️ Note that there were still cases where net sales are inconsistent between product and order reports, such as the issue #56096. In this PR we are focusing on fixing the product reports (`wp_wc_order_product_lookup` table).

Closes #32837 #48761 #48775.

#### Next steps

There is still an issue #48954 that is part of the epic issue #50480 need to be fixed.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Follow the steps in #48768 and #51262, the inconsistent net sales should not occur for partial and full refunds.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

octaedro committed:
Fix inconsistent behavior for Net Sales in Analytics for single line refund (#48768)

* Fix behavior for single line refund

* Add comments

* Add hook

* Change version trigger

* Add changelog

* Fix net_revenue

* Revert gross_revenue changes

* Update version to include the fix

* Update target release version to 9.4

---------

Co-authored-by: Ian Yu-Hsun Lin <ianlin@automattic.com>

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

ianlin committed:
Fix inconsistent behaviour for Net Sales in Analytics for full refund (#51262)

* Update primary key to composite (order_item_id, order_id) in the wc_order_product_lookup table

* Add full refunded order record to wc_order_product_lookup table

* Delete order items in order product lookup when the order is full refund

The previous commit will add the order item with negative net revenue
when the order is fully refunded. But the items sold in the products
analytics page is incorrect, because there was a DB record of an order item with
positive net revenue, and the product analytics query simply sums up all
the positive net revenue.

* Revert "Update primary key to composite (order_item_id, order_id) in the wc_order_product_lookup table"

This reverts commit 785a66521797ab14e84950b27a53f6ef941aacd7.

* Delete order item with negative sales recorded earlier due to partial refund

* Extract the logic of delete full refunded orders to a separate method

* Add changefile(s) from automation for the following project(s): woocommerce

* Revert "Extract the logic of delete full refunded orders to a separate method"

This reverts commit 47325f69d6f549d3c591270a877647594c403d9a.

* Revert "Delete order item with negative sales recorded earlier due to partial refund"

This reverts commit 18751d619364c2a5ad8ed1193da30a858eb04ea1.

* Revert "Delete order items in order product lookup when the order is full refund"

This reverts commit 1af8158e351b5f2bbe09796e2c1e367bf719eb04.

* Revert "Revert "Update primary key to composite (order_item_id, order_id) in the wc_order_product_lookup table""

This reverts commit b682c393bfccdb426fabf49e6704a5075d60f54b.

* Remove the partially refunded order item if exists

* Update analytics order stats test

Since we have added refunded order line items in wc_order_product_lookup
table, the unique product count should also relect the amount of
refunded products.

* Ignore refunded products in extended info for analytic orders report

---------

Co-authored-by: github-actions <github-actions@github.com>

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

mikkamp commented on plugins/woocommerce/src/Admin/API/Reports/Products/DataStore.php:
```diff
@@ -432,17 +432,51 @@ public static function sync_order_products( $order_id ) {
 
+		$is_full_refund = false;
+		// Order fully refunded won't have the order line items.
+		if ( 'shop_order_refund' === $order->get_type() && empty( $order_items ) ) {
```
I don't think this check is very accurate, since get_items by default only queries `line_item`, which doesn't include shipping/tax/etc... For example I can have the following scenario:

1. Order with 2 products purchased and a shipping fee
2. Refund one of the products (at this stage the partial refund will have a single line item) 
3. Second refund of just the shipping fee (here the default get_items will also be empty)

This means that it will now remove the entry that was refunded in 2 and create a product refund entry for both products when only one was refunded (similar will have if we skip step 2 and only refund shipping).

This is [the check in WC for determining if a refund order is partial or not](https://github.com/woocommerce/woocommerce/blob/9.4.3/plugins/woocommerce/includes/wc-order-functions.php#L673), do we want to replicate that check or hook into the actions and save order meta so we can determine if it's partial or not?

ianlin replied:
Thanks for pointing this and providing the info of determining partial refund in WC core. I missed the fee factor in the original implementation. I will see which way is better - copy the check or save the wp_wc_orders_meta by hooking into the action.

ianlin replied:
I made it to save the refund type in `wc_orders_meta` table in 8cf996b.

mikkamp commented on plugins/woocommerce/src/Admin/API/Reports/Products/DataStore.php:
```diff
@@ -432,17 +432,51 @@ public static function sync_order_products( $order_id ) {
+				WHERE 1 = 1
+					AND order_stats.parent_id = %d
+					AND order_stats.total_sales < 0
+			";
```
This query seems like it's missing some scenarios or mislabeled as it removes more than just the partial refunded rows. When I tested refunding just shipping this removed all entries in the `wp_wc_order_product_lookup` table for a particular order.

As it's linking any entries in the product lookup table to the refunded order, so it's not only catching partial refund lines but any product sales that have been made for the parent order. If it was actually a full refund that might be fine, because the code below re-adds the rows as each product being refunded. But since the code below also uses a `$wpdb->replace` it's not clear to me why we need to remove rows and then add them back again.

ianlin replied:
Thanks. Instead of removing them first, I changed it to select the already existing partial refunds in lookup table, then exclude them from the parent orders line items. 1d54d24

mikkamp commented on plugins/woocommerce/includes/wc-update-functions.php:
```diff
@@ -2937,3 +2937,44 @@ function wc_update_950_tracking_option_autoload() {
+ */
+function wc_update_960_update_primary_key_to_composite_in_order_product_lookup_table() {
+	global $wpdb;
+	$wpdb->query( "ALTER TABLE {$wpdb->prefix}wc_order_product_lookup DROP PRIMARY KEY, ADD PRIMARY KEY (order_item_id, order_id)" );
```
Can you clarify why this is needed? From my understanding the original reasoning was:

> because we will have multiple `order_item_id = 0`. Every time an order is refunded, we will insert a new line with `order_item_id = 0` and the `order_id`

However looking at the code below I see that for a full refund we fetch line items using `$parent_order->get_items()`. So each entry in the product lookup table still has a unique line item ID and an order ID.

ianlin replied:
Thanks for catching this. The previous implementation did not really need the composite primary key, that was a mistake when changing implementations during development. However, the new implementation right now is to also make the primary key as composite one: `(order_item_id, order_id, product_id)`. Please see the point 3 in the https://github.com/woocommerce/woocommerce/pull/53019#issuecomment-2536151879.

mikkamp commented on plugins/woocommerce/includes/wc-update-functions.php:
```diff
@@ -2937,3 +2937,44 @@ function wc_update_950_tracking_option_autoload() {
+        SELECT COUNT(*)
+        FROM {$wpdb->prefix}wc_order_product_lookup AS product_lookup
+        WHERE product_lookup.order_id = order_stats.parent_id
+          AND product_lookup.product_net_revenue = order_stats.total_sales
```
This is not super harmful as the consequence will be re-importing additional orders.

However the order `total_sales` will not be equal  to the `product_net_revenue` in quite a variety of cases. For example:
- If we have ordered multiple products in a single order
- If an order has additional fees such as shipping
- `total_sales` will include taxes and `product_net_revenue` will exclude it 

So most likely this is just going to catch all orders that have some kind of refund attached, since the rows where the two matched will not have been recorded previously.

That behaviour might be desired, but it might be better to restructure the query so we can determine what it should be including.

ianlin replied:
Thanks, I made a change to exclude specifically refunded shipping fees and refunded tax fees.

mikkamp commented on plugins/woocommerce/includes/wc-update-functions.php:
```diff
@@ -2937,3 +2937,44 @@ function wc_update_950_tracking_option_autoload() {
+          AND product_lookup.product_net_revenue = order_stats.total_sales
+    ) = 0";
+
+	// phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared -- No user input in the query, everything is hardcoded.
```
We can avoid this ignore if we include the hardcoded query without using a temp variable. PHPCS will see that it's hardcoded.

```php
	$orders = $wpdb->get_results(
		"SELECT order_stats.order_id FROM {$wpdb->prefix}wc_order_stats AS order_stats
		WHERE order_stats.total_sales < 0
		AND (
			SELECT COUNT(*)
			FROM {$wpdb->prefix}wc_order_product_lookup AS product_lookup
			WHERE product_lookup.order_id = order_stats.parent_id
				AND product_lookup.product_net_revenue = order_stats.total_sales
		) = 0"
	);
```

mikkamp reviewed (commented):
When testing I originally came across the issue that Product reports are always based on the date_created, whereas for Order reports this can be date_completed (as specified in the settings). In my case the completed date was a day later so I had to select both days to compare reports.

Considering that the linked investigation mentions:
> These issues are complex, difficult to tackle, and challenging to test all the side effects.

Why don't we see more testing where at least the basic scenarios are covered. For example `WC_Admin_Tests_Reports_Products::test_populate_and_refund` only covers a partial refund in the test, where we are explicitly changing the behaviour for a full refund. A scenario where only a shipping fee is refunded could also be helpful.

Based on the comments I added, I don't feel this is fully ready for pushing to release. I'd feel a lot more confident if some of the scenarios I described are covered in unit tests so we are at least working with predictable outcomes.

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

ianlin commented:
@mikkamp Thanks for the review and for catching those issues!

I set up a new local site and generated 10 orders using Smooth Generator. Along with the day_created / day_paid issue you mentioned, I also noticed an inconsistency between Net Sales in the Products report and Net Sales in the Orders report. It turns out this inconsistency is caused by fees - fees aren't included in the Products report since the `wp_wc_order_product_lookup` table doesn't map fees to products. This behaviour also exists in the trunk branch. It feels logical for Net Sales to exclude fees, so the issue seems to be with the Orders report instead.

That said, I agree these inconsistencies need a closer review and more tests to address them properly. I'll mark this PR as a draft and let the HEs know it won't make it into the 9.6 release.

mikkamp commented:
> I also noticed an inconsistency between Net Sales in the Products report and Net Sales in the Orders report. It turns out this inconsistency is caused by fees - fees aren't included in the Products report since the wp_wc_order_product_lookup table doesn't map fees to products. This behaviour also exists in the trunk branch. It feels logical for Net Sales to exclude fees, so the issue seems to be with the Orders report instead.

Fees would be another difference to take into consideration. However my expectation would be that fees are not included in the products report, but are in the orders report. If that needs to be highlighted then maybe a tooltip or documentation would be the right place to do so. I wouldn't consider that an issue for the orders report.

In my opinion this PR should solely focus on the refunds, and ensure any product refunds are reflected in the `wp_wc_order_product_lookup`. So they don't continue to count as sales in some cases.

ianlin committed:
Hook refunded actions and add refund type to orders meta

ianlin committed:
Update primary key to composite (order_item_id, order_id, product_id) in the wc_order_product_lookup table

ianlin committed:
Get the parent order items when it's a full refund, and exclude the
already partially refunded items.

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

ianlin committed:
Update order stats report tests by adding a line item in the refund

ianlin committed:
Add tests for full refunds and shipping fee refunds in product reports

ianlin committed:
Update primary key to composite (order_item_id, order_id, product_id) in the wc_order_product_lookup table

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

ianlin committed:
Update migration SQL for refunded orders not presented in product loolup
table

ianlin commented:
@mikkamp Thanks again for catching those issues and providing potential solutions.

I've updated the code to reflect all the suggestions you gave. Here's the notable changes:

1. Hook refunded actions to add refund type in `wp_wc_orders_meta` table. This resolve the issue that refunded shipping fee also does not have line items - now the refunded shipping fee is marked as partial refund thanks to the hooks.
2. For the case of full refund, there was a case that the order was refunded partially first then refunded fully after. Since the full refund does not have order line items,  my original implementation was to remove the partial refunds that exists in product look table, and use the parent order's order line items to update to the product lookup table. The new implementation is, select those already existing partial refunds in lookup table first, and exclude them from the parent orders line items.
3. Each of the refunded line items in a full refund is written to the product lookup table, with the `order_item_id` set to `0` and the `order_id` set to the refunded order ID. As a result I made the primary key to a composite key as `(order_item_id, order_id, product_id)`, since there will be multiple items with the same `order_item_id` as `0` and the same `order_id` for the refunded order.
4. Update the test for order stats, since I found there's one test that did not add line items for the partial refund, which makes the test result inaccurate.
5. Add tests for full refunds in product reports
6. Update the migration SQL by specifically exclude refunded shipping fees and refunded tax fees

I agree those issues are really complex, and I've tried my best to explain the changes above, but please let me know if anything is unclear.

mikkamp commented on plugins/woocommerce/includes/wc-update-functions.php:
```diff
@@ -2932,3 +2932,49 @@ function wc_update_950_tracking_option_autoload() {
+				SELECT COUNT(*)
+				FROM wp_wc_order_product_lookup AS product_lookup
+				WHERE product_lookup.order_id = order_stats.order_id
+					AND product_lookup.product_net_revenue = order_stats.total_sales
```
This is still not going to match if the refund order refunded two products at once. But considering that this is a run once scenario, and ideally it shouldn't have any refund lines in the product_lookup table before this change. It's not going to make a difference. And if it is run a second time then it will just catch extra orders to recalculate.

ianlin replied:
> ideally it shouldn't have any refund lines in the product_lookup table before this change

Agreed, I've removed this check, since it's not possible that refund items exist before this fix.

mikkamp commented on plugins/woocommerce/src/Admin/API/Reports/Products/DataStore.php:
```diff
@@ -432,17 +466,79 @@ public static function sync_order_products( $order_id ) {
+					return $refunded_product_id['product_id'];
+				},
+				$refunded_product_ids
+			);
```
If we aren't worried about type casting, we can use `array_column` for this. Or alternatively if we use `wp_list_pluck` then we can work with a list of objects and we don't need to change the type to `ARRAY_A` for the DB call.

```suggestion
			$refunded_product_ids = array_column( $refunded_product_ids, 'product_id' );
```

mikkamp commented on plugins/woocommerce/src/Admin/API/Reports/Products/DataStore.php:
```diff
@@ -432,17 +466,79 @@ public static function sync_order_products( $order_id ) {
+					$parent_order_items,
+					function ( $order_item ) use ( $refunded_product_ids ) {
+						$order_item_id = $order_item->get_id();
+						$product_id    = wc_get_order_item_meta( $order_item_id, '_product_id' );
```
We should only be interested in [WC_Order_Item_Product](https://woocommerce.github.io/code-reference/classes/WC-Order-Item-Product.html) here, if so it has a `get_product_id` function. Although that points out a flaw in this logic.

I can purchase two separate variations from the same product. That's going to give me two line items with an identical `product_id`. Does that mean we should be expanding the primary key to include `variation_id`?

I'm still not totally understanding why we need such a liberal key and why `order_item_id` needs to forced to 0. See comment below where we set it to 0.

ianlin replied:
Thanks for pointing the variation product issue, updated in [5566b51](https://github.com/woocommerce/woocommerce/pull/53019/commits/5566b51e2ff9d16cbc9eba40f9a7ccb6497c53f1).

mikkamp commented on plugins/woocommerce/src/Admin/API/Reports/Products/DataStore.php:
```diff
@@ -432,17 +466,79 @@ public static function sync_order_products( $order_id ) {
+			);
+
+			// Exclude the refunded product ids from the parent order items.
+			$order_items = array_values(
```
If we are not concerned about the array index then this seems a little unnecessary, and we can just work with the array returned from `array_filter`.

mikkamp commented on plugins/woocommerce/src/Admin/API/Reports/Products/DataStore.php:
```diff
@@ -432,17 +466,79 @@ public static function sync_order_products( $order_id ) {
+				WHERE 1 = 1
+					AND order_stats.parent_id = %d
+					AND order_stats.total_sales < 0
+			";
```
I'm not so sure this kind of filtering is going to work in all cases. Because this means for a full refund we are going to skip any line item with a matching product (variations should be taken into consideration here).

However if I have a partial refund order where I decided to do a 40% refund on a particular product and then later decided to fully refund the order, we are detecting this product as "partially refunded", so we'll skip it. Which means in the reports it will still be counting 60% of the product towards sales.

Also because a line item (and a row in the `product_lookup` table) has a `product_qty`. If I have an order where multiples of a product are purchased and only one is refunded, and later the whole order, then we again end up with the same problem.

I think in this case it would be best to calculate the difference with all found partial refunds and then add that as a new product_lookup row to indicate the remaining refund. A simpler method would be to remove old entries, but I have the feeling that would be incorrect. Because each entry in product_lookup has a date timestamp, so that would attribute the refund to the wrong time. So what we really want to see is the full history, ex:

- 2024-12-05 Product ordered $20 revenue
- 2024-12-06 Partial refund -$5 (total revenue for that product is still $15)
- 2024-12-07 Full refund -$15 (no more revenue)

If I view the report for all of December then I'll see no revenue. But if I view breakdown reports, then I can still see the accurate amount of positive or negative revenue that occurred on those days.

ianlin replied:
Thanks for the reminder. Updated in da6ad3e and a88e12d.

mikkamp commented on plugins/woocommerce/src/Admin/API/Reports/Products/DataStore.php:
```diff
@@ -455,15 +551,15 @@ public static function sync_order_products( $order_id ) {
-					'order_item_id'         => $order_item_id,
+					// When the order is full refund, there is no order items, so the order_item_id should be 0.
+					'order_item_id'         => $is_full_refund ? 0 : $order_item_id,
 					'order_id'              => $order->get_id(),
```
For the order ID we are using the refund order, but the line item ID's are unique entries from the parent order for full refunds and from the refund order for a partial refund.

Since that will still make the order item ID unique in combination with the order number, why can't we just always populate it? Which means we don't need to expand the primary key across product_id (and variation_id).
Does this bring us into trouble when we populate the reports where it might try to fetch/populate the line item details directly linked to the refund order?

I tried adding an entry like this where I use the refund order ID but the order_item_id of the parent product. In the report it seems to populate the data from the product directly, so I don't see anything incorrect:
![image](https://github.com/user-attachments/assets/b5c3f259-33a6-4961-8701-f2f69c1e4531)

![image](https://github.com/user-attachments/assets/e0733077-21a9-4b2a-8f00-625bc2079607)

ianlin replied:
> Since that will still make the order item ID unique in combination with the order number, why can't we just always populate it?

You're absolutely right. Not sure how did I get that far expanding the primary key to `product_id`. 😓 Updated in 7c85581.

mikkamp commented on plugins/woocommerce/src/Admin/API/Reports/Products/DataStore.php:
```diff
@@ -111,6 +111,40 @@ protected function assign_report_columns() {
 	 */
 	public static function init() {
 		add_action( 'woocommerce_analytics_delete_order_stats', array( __CLASS__, 'sync_on_order_delete' ), 10 );
+		add_action( 'woocommerce_order_partially_refunded', array( __CLASS__, 'add_partial_refund_type_meta' ), 10, 2 );
+		add_action( 'woocommerce_order_fully_refunded', array( __CLASS__, 'add_full_refund_type_meta' ), 10, 2 );
```
Should we be concerned that this won't work for refunds created before the fix? Even if we go to re-import orders in Analytics > Settings, it won't find the meta value and default them to partial_refunds.

ianlin replied:
Thanks for the reminder, I've added in 5e1bde3. Since we only care about the full refunds without line items, I added the `_refund_type` meta to orders without line items found in the query from the migration script.

mikkamp commented on plugins/woocommerce/src/Admin/API/Reports/Products/DataStore.php:
```diff
@@ -432,17 +466,79 @@ public static function sync_order_products( $order_id ) {
+		$is_full_refund = false;
+		$refund_type    = $order->get_meta( '_refund_type' );
+		// When the order is a full refund, there is no order items.
+		// We need to get the parent order items, and exclude the items that is already being patially refunded.
```
Can you clarify how it was determined that there are no line items for a full refund? Am I doing something wrong when issuing a full refund? I got one of my test orders and made sure to include everything for a refund (this order had no prior partial refunds):

![image](https://github.com/user-attachments/assets/56168dca-bb81-4a1b-a63e-9a6fa6eed0d8)

Now if I look up order meta for the refund order I can see it detected it as a full refund:
![image](https://github.com/user-attachments/assets/59470ad5-ef03-4791-93f1-0756fc82e40f)

When I check the line items for the refund order, it has one for each product item:
![image](https://github.com/user-attachments/assets/3b917033-6f80-4835-92cb-e002e2f0ee6a)

Also not finding any other method to refund which might generate the refund order without line items when looking through the [instructions here](https://woocommerce.com/document/woocommerce-refunds/).

Update: I did some searching, you can also toggle the status of an existing order to refunded. Which will create a refund order with the total amount and no line items. I believe that scenario is not very common though, because it means you can't do an automatic refund using the payment gateway. So this will probably be common only for stores that handle payment outside of WooCommerce.

I guess that means we should still include the empty line items check and not fully rely on the $refund_type, as in some cases there will be line items.

ianlin replied:
Thanks, and yeah, I was only handling full refunds created by changing the order status to "Refunded", and I didn't realise that refunding all the products one by one in an order would also count as a "full" refund. I've added back the check for empty order line items.

mikkamp commented on plugins/woocommerce/src/Admin/API/Reports/Products/DataStore.php:
```diff
@@ -432,17 +466,79 @@ public static function sync_order_products( $order_id ) {
+			// So the net revenue and product quantity should be negative.
+			if ( $is_full_refund ) {
+				$net_revenue = -abs( $net_revenue );
+				$product_qty = -abs( $product_qty );
```
Should this be limited to only quantity and revenue? What about taxes, shouldn't that also be reversed? And this seems to indicate we can also have per product shipping, or coupon amounts per product? Shouldn't that be marked as fully refunded as well? Otherwise the gross revenue calculation is going to be incorrect since it's done with:

`'product_gross_revenue' => $net_revenue + $tax_amount + $shipping_amount + $shipping_tax_amount`

mikkamp reviewed (commented):
Thanks for adding the additional testing that's helpful to cover more scenarios.

I tried to be very verbose with the comments as I tried to wrap my head around how all the sections work. So sorry if there are a lot of comments in there.

However considering that variations aren't taken into consideration and that in some circumstances full refunds can contain line items, I don't feel comfortable to approve this yet.

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

ianlin committed:
Check for empty order items when full refund is triggered by changing
order status to "Refunded"

ianlin committed:
Change the primary composite key back to (order_item_id, order_id) and
do not store 0 for order_item_id when the full refund does not have
order line items

Because in this case although the order_item_id will be repeated but order_id
will be the refund order ID, the combination of them will still be
unique.

ianlin committed:
Handle partially refunded products

If a single line item was refunded 60% then fully refunded after, we need store the difference in the product lookup table.
E.g. A product costs $100, it was previously partially refunded $60, then fully refunded $40.
So it will be -abs( 100 + (-60) ) = -40.;

ianlin committed:
Handle the case when there are multiple partially refunded products in
an order

ianlin committed:
Handle variation products for full refund without line items

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

ianlin committed:
Fix PHPCS

ianlin committed:
Use %i as the placeholder for the table name in the prepare query

The prepare placeholder "%i" was released in WP 6.3, and the current minimum WP version for WC is 6.5.

ianlin committed:
Handle the old refund orders don't have _refund_type metadata in
migration script

ianlin committed:
Remove condition checking for refunded items in product lookup table in
migration script

ianlin committed:
Update migration scripts version to WC 9.7

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

ianlin committed:
Add a test for when all line items in an order are refunded each by 100%

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

ianlin commented:
@mikkamp Thanks for the detailed testing and providing more scenarios that I did not think of. I updated the code to address most of your concerns. I implemented it based on the expected results you mentioned:

> So what we really want to see is the full history, ex:
>
> 2024-12-05 Product ordered $20 revenue
> 2024-12-06 Partial refund -$5 (total revenue for that product is still $15)
> 2024-12-07 Full refund -$15 (no more revenue)

I also handle when the two variations with the same parent product were refunded case.

Just one thing I wasn't sure how to handle:

> Should this be limited to only quantity and revenue? What about taxes, shouldn't that also be reversed? And this seems to indicate we can also have per product shipping, or coupon amounts per product? Shouldn't that be marked as fully refunded as well? Otherwise the gross revenue calculation is going to be incorrect since it's done with:
>
> 'product_gross_revenue' => $net_revenue + $tax_amount + $shipping_amount + $shipping_tax_amount

This is the case when the order refund type is `full`, and the refunded order does not have line items (i.e. order status changed to `Refunded`). In this case the `$order_items` is the parent order's line items, and the `$order` here is still the refunded order, then the shipping amount (`$order->get_item_shipping_amount( $order_item )`) returns nothing. I think since this PR specifically solves the inconsistency in net revenue, and I don't see product gross revenue is displayed in the Products report, we could probably solve the shipping amount (and tax, coupon, etc) in the follow up PR, as I haven't thought of a good way to handle it?

mikkamp commented on plugins/woocommerce/src/Admin/API/Reports/Products/DataStore.php:
```diff
@@ -432,17 +466,144 @@ public static function sync_order_products( $order_id ) {
+
+			// Get the partially refunded products ID and revenue from the parent order.
+			$partial_refund_products = $wpdb->get_results(
+				$wpdb->prepare(
```
We can also opt to do a `GROUP BY product_id, variation_id` and then get the sum of revenue. That would simplify the array_reduce or array_walk to only separating variation_ids.

Also a `variation_id` would not be identical to `product_id`, so do we really need to keep them separate in the lookup table? If we create one lookup array indexed by either `variation_id` or `product_id` then we wouldn't need to set both key and id later on when we want to lookup a value.

mikkamp commented on plugins/woocommerce/src/Admin/API/Reports/Products/DataStore.php:
```diff
@@ -432,17 +466,144 @@ public static function sync_order_products( $order_id ) {
+			};
+
+			// Filter the items that is already being patially refunded from the parent order item.
+			$order_items = array_filter(
```
From my understanding the reason we are pre-filtering the order items here is because we could have some items which have already been refunded fully. However this means that we have two locations where we calculate the previous refund amount per line item. Can't we simplify that and have it go through all the parent line items, and just do a check if the refund amount is 0 then skip to the next line item.

If we simplify that, it will also make it easier to support cases where all the revenue except taxes was refunded.

So in this part I'm suggesting to completely remove the filtering of $order_items and it's helper function. And then a little further down if `$is_full_refund_without_line_items` is true and `$net_revenue` is zero we can just `continue` and skip to the next line item.

mikkamp commented on plugins/woocommerce/src/Admin/API/Reports/Products/DataStore.php:
```diff
@@ -432,17 +466,144 @@ public static function sync_order_products( $order_id ) {
+					$net_revenue = -abs( $net_revenue + $partial_refund_product_revenue[ $key ][ $id ] );
+				} else {
+					$net_revenue = -abs( $net_revenue );
+				}
```
If the previous two suggestions to simplify are taken into account then we can reduce this code a bit more.

```suggestion
			if ( $is_full_refund_without_line_items ) {
				$id             = $variation_id ?: $product_id;
				$partial_refund = $partial_refund_product_revenue[ $id ] ?? 0;
				$net_revenue    = -abs( $net_revenue + $partial_refund );
				
				// Skip items that have already been fully refunded (single or multiple partial refunds).
				if ( 0.0 === $net_revenue ) {
					continue;
				}
```

mikkamp reviewed (commented):
Thanks for all the refactoring, I know it's gone through a lot of changes. But when looking at the final combination of code, it looks like we could simplify some points a bit more. I've added separate comments for this.

> I think since this PR specifically solves the inconsistency in net revenue, and I don't see product gross revenue is displayed in the Products report, we could probably solve the shipping amount (and tax, coupon, etc) in the follow up PR, as I haven't thought of a good way to handle it?

While I do agree that we can address it in a follow up PR, the thing I'm hesitant about is that this functionality will land in the next release before we have the chance for a follow up PR. Since we've included a DB update function it will go through and fix entries for previous refund orders. So I think we should try and avoid a scenario where we have one release fix entries for most orders and then have the next release which again needs to fix the remaining entries.

I would prefer if we can simplify this code as much as possible (as suggested in the comments), and then add a few more test scenarios where we take additional things like taxes and coupons etc. into account. I'd personally go for a test driven development here and first create some extra unit tests and then tweak the code to make the test pass.

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

ianlin committed:
Simplify the code for the case of full refunds without line items

ianlin committed:
Calculate shipping amount when the order status chanced to REFUNDED

i.e. A full refund without order line items.

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

ianlin committed:
Handle the case when refund an order item then do a full refund

The negative shipping fee should be distributed based on the number of remaining
items left to refund, rather than the number of total order items.

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

mikkamp commented on plugins/woocommerce/src/Admin/API/Reports/Products/DataStore.php:
```diff
@@ -433,17 +467,106 @@ public static function sync_order_products( $order_id ) {
+			// If the order is a full refund and there is no order items. The order item here is the parent order item.
+			if ( $is_full_refund_without_line_items ) {
+				$id             = $variation_id ? $variation_id : $product_id;
+				$partial_refund = $partial_refund_product_revenue[ $id ];
```
This needs a fallback to set to 0 as there might not be a partial refund for this product. It's causing the [unit tests to fail](https://github.com/woocommerce/woocommerce/actions/runs/12903128065/job/35977882888?pr=53019).

This was my previous suggestion:
```suggestion
				$partial_refund = $partial_refund_product_revenue[ $id ] ?? 0;
```

ianlin committed:
Add a fallback to set to 0 for the partial_refund_product_revenue lookup

ianlin committed:
Calculate shipping tax amount when the order status changed to REFUNDED

i.e. A full refund without order line items.

ianlin committed:
Fix wpdb prepare for IN clause for OrdersTableDataStore::get_total_tax_refunded

ianlin committed:
Calculate cart tax amount when the order status changed to REFUNDED

i.e. A full refund without order line items.

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

ianlin committed:
Fix PHPCS

ianlin committed:
Set coupon_amount to 0 when the order status changed to REFUNDED

i.e. A full refund without order line items.

ianlin commented:
@mikkamp Thanks again for all the reviews and suggestions. I've added more commits that make the value of `product_gross_revenue` correct when it's a full refund without line items by calculating the `tax_amount`, `shipping_amount`, `shipping_tax_amount`. Note that I set the coupon amount to `0` in that case because it's aligned with the behaviour when partially refund 1 product from the order.

I will update the migration script's versioning when the PR is ready for merge.

mikkamp commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:
```diff
@@ -1047,8 +1050,12 @@ public function get_total_tax_refunded( $order ) {
+					$order->get_id(),
+					$include_tax_amount ? 'tax_amount' : '',
+					$include_shipping_tax_amount ? 'shipping_tax_amount' : '',
+				)
```
Not wrong but was initially confusing as I think most use the same format as `sprintf`. [According to the docs](https://developer.wordpress.org/reference/classes/wpdb/prepare/) both are allowed:

> Arguments may be passed as individual arguments to the method, or as a single array containing all arguments. A combination of the two is not supported.

mikkamp commented on plugins/woocommerce/includes/class-wc-order.php:
```diff
@@ -2161,18 +2161,21 @@ public function get_total_refunded() {
 		}
 
-		$total_refunded = $this->data_store->get_total_tax_refunded( $this );
+		$total_refunded = $this->data_store->get_total_tax_refunded( $this, $include_tax_amount, $include_shipping_tax_amount );
```
The two booleans have only been added to the datastore for HPOS, shouldn't it also be added to `WC_Order_Data_Store_CPT` and `WC_Order_Data_Store_Interface` as well?

mikkamp commented on plugins/woocommerce/includes/class-wc-order.php:
```diff
@@ -2161,18 +2161,21 @@ public function get_total_refunded() {
 	 * @return float
 	 */
-	public function get_total_tax_refunded() {
+	public function get_total_tax_refunded( $include_tax_amount = true, $include_shipping_tax_amount = true ) {
```
Can you clarify how this handles the cache key? As it is now it looks risky, because if I call this two times in a row the same cache key would be used:

1. `get_total_tax_refunded()` - All order taxes
2. `get_total_tax_refunded( false, true )` - Only shipping taxes

I haven't tested, but based on the logic I'd assume both would return the same value.

mikkamp commented on plugins/woocommerce/src/Admin/API/Reports/Products/DataStore.php:
```diff
@@ -433,30 +467,126 @@ public static function sync_order_products( $order_id ) {
+
+					// Calculate the shipping tax amount to refund from the parent order.
+					$shipping_tax                 = (float) $parent_order->get_shipping_tax();
+					$total_shipping_tax_refunded  = $parent_order->get_total_tax_refunded( false, true );
```
Do we ever call this to fetch the total taxes refunded without shipping taxes? Because if the only case is to get the total shipping taxes, then I think it would be safer to create a separate function called `get_total_shipping_tax_refunded`. As mentioned in other comments, that would make it easier to set a unique cache key and would reduce the risk of having to change existing methods.

It would need to be implemented for each data store which would be a risk in itself in case there are plugins that extend/change the datastore. But maybe we can safeguard against that by checking if the datastore has the function `get_total_shipping_tax_refunded`.

ianlin replied:
Thanks for the suggestion, I wasn't aware of the cache and it makes 100% sense to create a separate method for total shipping tax refunded.

> we can safeguard against that by checking if the datastore has the function get_total_shipping_tax_refunded

I'm not sure what should be doing when the method of the datastore object does not exist. Returning `0` sounds not right as the result would be incorrect. Maybe throws an error?

mikkamp replied:
I'd be hesitant to error out here, from my understanding that would block any of the rows being recorded for a specific refund order.

I think we should be ok with leaving it at 0 in that case (log a non blocking warning if possible). It would then still calculate the rest just without an accurate shipping tax refunded rate (since the datastore didn't support it).

Looking at the CPT implementation for `get_total_shipping_tax_refunded` we default to 0 anyways if the query fails: `?? 0`
So might as well default to the same if the function doesn't exist in the datastore.

mikkamp commented on plugins/woocommerce/src/Admin/Overrides/OrderTraits.php:
```diff
@@ -42,29 +44,35 @@ public function get_item_shipping_amount( $item ) {
-					$total_shipping_tax_amount += $tax_item_total;
+		$total_shipping_tax_amount = $shipping_tax_amount ? $shipping_tax_amount : 0;
+
+		if ( 0 === $total_shipping_tax_amount ) {
```
I know we check against 0 before we call the function. But wouldn't it be safer to check the passed shipping_tax_amount to stay closer to the original implmentation?

```suggestion
		if ( null === $shipping_tax_amount ) {
```

mikkamp commented on plugins/woocommerce/src/Admin/Overrides/OrderTraits.php:
```diff
@@ -42,29 +44,35 @@ public function get_item_shipping_amount( $item ) {
 			}
 		}
+
 		return $total_shipping_tax_amount / $order_items * $product_qty;
```
From what I can see we override this function just to do this one line of calculation. Since we already passed it `$total_shipping_tax_amount` and `$order_items`, wouldn't it be safer to just make a helper function within `Reports/Products/DataStore` to do this calculation? It seems like it's a bit of a niche calculation we are doing here and it's not clear to me how adding this to the OrderTraits makes it useful for others to use (same applies for `get_item_shipping_amount`).

mikkamp replied:
I checked again and it seems OrderTraits are only added to Orders classes used in WC Admin, so in that aspect it will be safer to change their behaviour. However I'd still add some more context in which cases we might want to override the amounts, by describing a scenario with partial refunds in an order.

mikkamp commented on plugins/woocommerce/src/Admin/Overrides/OrderTraits.php:
```diff
@@ -42,29 +44,35 @@ public function get_item_shipping_amount( $item ) {
 
 		$product_qty               = $item->get_quantity( 'edit' );
 		$order_taxes               = $this->get_taxes();
 		$line_items_shipping       = $this->get_items( 'shipping' );
```
These two lines are only needed if we are doing the foreach loop, so I suggest we move them within the if condition.

mikkamp commented on plugins/woocommerce/src/Admin/Overrides/OrderTraits.php:
```diff
@@ -80,4 +88,24 @@ public function get_item_shipping_tax_amount( $item ) {
+	 *
+	 * @param WC_Order_Item $item Line item from order.
+	 *
+	 * @return float|int
```
I guess this happens in other functions in this class, but if we initialize as `0.0` then we can always return a float.

mikkamp commented on plugins/woocommerce/src/Admin/Overrides/OrderTraits.php:
```diff
@@ -80,4 +88,24 @@ public function get_item_shipping_tax_amount( $item ) {
+
+		foreach ( $order_taxes as $tax_item ) {
+			$tax_item_id      = $tax_item->get_rate_id();
+			$cart_tax_amount += isset( $tax_data['total'][ $tax_item_id ] ) ? (float) $tax_data['total'][ $tax_item_id ] : 0;
```
Can you clarify a bit more about what's being calculated here? From what I can see it's only adding up the tax rates if those same ID's are found in the order taxes.

In what cases do the tax ID's not line up with tax ID's in the order? For example in which scenarios can we not add up just all the taxes recorded for this line item?

ianlin replied:
It was actually copied from the original implementation in `Reports/Products/DataStore` here:
https://github.com/woocommerce/woocommerce/pull/53019/files#diff-d2438118f148f28d98ab96e1a12bdd753e462f4659970939f2fe3e2b87155f59L450-L456

And I think since an order line item would only have one tax rate (is it true?), we can just assign the rate to `$cart_tax_amount` when match. The commit is 40e9f29.

mikkamp replied:
With compound taxes, it can get complicated:
![image](https://github.com/user-attachments/assets/7c400451-4da7-46b3-8c43-4ed4a1443255)

As we can see now in the order we have both tax rates applied per line item (also for shipping since the tax rate is set for shipping):
![image](https://github.com/user-attachments/assets/8ca969a6-0fe6-4f9c-afe6-74da9bdf1c86)

It might not be common to configure this manually but there are automatic tax plugins that apply scenarios where they might have a state tax and a city tax. So we do have to take it into consideration.

mikkamp replied:
> It was actually copied from the original implementation in `Reports/Products/DataStore`

I can't think of a scenario where total order taxes doesn't include line item taxes, as generally we'd expect the order totals to be recalculated as soon as something changes. But I suppose since this code has been around for  along time we could leave it as is, as it still calculates the same thing.

mikkamp reviewed (commented):
Thanks for the changes so far, that's looking great. I left some comments though and think I'd need some further explanations to make sure we are covering edge cases. Some tax scenarios can get really complicated, so want to make sure we aren't introducing inconsistencies.

ianlin committed:
Create a function get_total_shipping_tax_refunded to ensure the cache is
different from get_total_tax_refunded

ianlin committed:
Add a test for the function get_total_shipping_tax_refunded()

ianlin committed:
Check the passed argument instead of the one init by it

ianlin committed:
Move the lines within the if condition as it's only used in it

ianlin committed:
Makes the method get_item_cart_tax_amount() return float consistently

ianlin committed:
Add comments in OrderTraits for the scenario of partial refunds in an
order

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

ianlin committed:
Refactor: get cart tax amount when it's set instead of adding all up

Since an order line item has only one tax rate.

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

mikkamp commented on plugins/woocommerce/includes/class-wc-order.php:
```diff
@@ -2161,27 +2161,45 @@ public function get_total_refunded() {
+			return $cached_data;
+		}
+
+		$total_shipping_tax_refunded = $this->data_store->get_total_shipping_tax_refunded( $this );
```
As mentioned I'd add the fallback to 0 here if the function doesn't exist in the data store.

mikkamp commented on plugins/woocommerce/src/Admin/Overrides/OrderTraits.php:
```diff
@@ -118,8 +118,11 @@ public function get_item_cart_tax_amount( $item ) {
+			$tax_item_id = $tax_item->get_rate_id();
+			if ( isset( $tax_data['total'][ $tax_item_id ] ) ) {
+				$cart_tax_amount = (float) $tax_data['total'][ $tax_item_id ];
+				break;
```
As mentioned in the other comment about compound taxes, I think it's safer to leave this as a `+=` and remove the break.

mikkamp reviewed (commented):
Thanks for all the changes I think we are almost there now.

If we can just get a fallback for the new function not existing in the datastore. As well as a tweak to support compound taxes, and maybe an extra unit test to make sure we get expected results.

After that I think we should be pretty much there.

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

ianlin committed:
Fallback to return 0 when the method get_total_shipping_tax_refunded
doesn't exist in the datastore

ianlin committed:
Revert "Refactor: get cart tax amount when it's set instead of adding all up"

This reverts commit 40e9f294d8e7cb50d0d8ba6e767ba7b7ba12116a.

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

ianlin committed:
Add a test for full refund compound tax in analytics sync orders

ianlin committed:
Merge branch 'trunk' into fix/32837-48761-48775-48954-inconsistent-behavior-of-net-sales-in-analytics

ianlin committed:
Rename DB migration functions so they are targeting WC 9.9

ianlin commented:
@mikkamp Thanks again for the suggestions. After testing, it looks like the compound tax case is already covered. I went ahead and added a test for a full refund with compound tax just to be sure.

mikkamp commented on plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/reports/class-wc-tests-reports-products.php:
```diff
@@ -491,4 +696,382 @@ public function test_report_export_arguments() {
+		WC_Helper_Queue::run_all_pending( 'wc-admin-data' );
+
+		// Get the last order item from the order.
+		$order_item_id = end( $order->get_items() )->get_id();
```
I get an error when trying to run this test locally:

> Only variables should be passed by reference

```suggestion
		$order_items   = $order->get_items();
		$order_item_id = end( $order_items )->get_id();
```

I understand that in GitHub actions it's running with `wp-env` which is different then I run it. Although it's not clear to me why the same error wouldn't be reported there. I tried testing with different PHP versions and I get the same error all the way from 7.4 - 8.3

There are a few locations in this file where we call `end` on the response of a method call.

mikkamp commented on plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/reports/class-wc-tests-reports-products.php:
```diff
@@ -491,4 +696,382 @@ public function test_report_export_arguments() {
+
+		$result = $wpdb->get_results(
+			$wpdb->prepare(
+				'SELECT * FROM wp_wc_order_product_lookup WHERE order_item_id = %d AND order_id = %d',
```
Like I mentioned I'm not running the tests through `wp-env`, so when I run them the WP DB prefix is `wptests_` which causes this to fail because the prefix `wp_` is hardcoded.

```suggestion
				"SELECT * FROM {$wpdb->prefix}wc_order_product_lookup WHERE order_item_id = %d AND order_id = %d",
```

There are a few locations in this file where this happens.

mikkamp reviewed (commented):
Thanks for all the changes. I ran the unit tests and just found a few minor adjustments needed to allow them to run in all environments.

I also did some testing on my site, but I'd like to clarify what we are expecting to fix with these issues. I see they mostly mention:
> Net sales total displayed in the Products report doesn't match the total we show in the Orders report

So I tested fully refunding an order. The products table is now showing the correct amounts:
![image](https://github.com/user-attachments/assets/f3ae3eff-b159-411c-b5c6-ab43cbc1447e)

Which is reflected in the report that there have been 0 sales:
![image](https://github.com/user-attachments/assets/95df1935-1932-4bb0-9551-a71f29e663ef)

But then when I look at the order stats table I get a different picture:
![image](https://github.com/user-attachments/assets/815eaf60-a304-4ad5-b697-3e31aea3687e)

Which shows as a negative net sales for both the Orders and the Revenue reports:
![image](https://github.com/user-attachments/assets/0b0a6cb1-ae5d-40c9-8116-bf2f712d13ea)

![image](https://github.com/user-attachments/assets/085e5c75-01e2-48fb-8703-77b5b21ccbb0)

Or do we attribute that to a flaw in the core order handling where if the status is changed to fully refunded it attributes the full refund amount to just a total:
![image](https://github.com/user-attachments/assets/ce470054-5901-4204-83a1-900e7181d76d)

When I refund per line item the entries in the order_stats table look better:
![image](https://github.com/user-attachments/assets/b6e9b1b9-5982-478e-9e9d-bb4ff6ac6cfb)

But the totals in the order and revenue reports are still mixed, the products report is showing correctly, so at least we did fix that.

Are we leaving the other two reports out of the picture? Because if we do we might need to update some of the changelog entries / PR description to clarify what changes we are covering.

ianlin committed:
Pass a variable to end() function instead of a function call

ianlin committed:
Use $wpdb->prefix in the query for unit tests

ianlin commented:
> Or do we attribute that to a flaw in the core order handling where if the status is changed to fully refunded it attributes the full refund amount to just a total:

Thanks for catching that. I believe it's a bug in the order reports, created an issue #56096.

I will update the changelog and the PR description.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

mikkamp reviewed (approved):
Thanks for the final changes. This looks ready now, so I'll go ahead and approve it.

I've just got two quick remaining comments:
1. The ` @since` tags are still pointing to 9.8.0 (do we have a placeholder?)
2. There are three changelog entries included with this PR, can we clarify in all of them that we are only fixing product reports (or merge the changelog entries)

ianlin committed:
Update @since tags to 9.9.0

ianlin committed:
Merge the changelogs to clarify the fix only focuses on product reports

ianlin commented:
Thanks for approval!

I've merged the three changelogs into one so it's clearer that we're only solving the issues in product reports.

For the `@since` placeholder like `x.x.x`, as far as I recall WC core doesn't support it and will cause the linter error. I've changed the version for `@since` to `9.9.0`, thanks for catching it.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

ianlin committed:
Remove the auto-generated changelog
